src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java

```diff
@@ -227,14 +242,58 @@ int getTimeout() {
                 MqttClient.DEFAULT_MQTT_OPERATION_TIMEOUT, MqttClient.MQTT_OPERATION_TIMEOUT_KEY));
     }
 
-    private void resubscribe() {
-        subscriptionTopics.forEach((key, value) -> {
-            subscribe(key, value).whenComplete((i, t) -> {
-                if (t != null) {
-                    logger.atError().kv(TOPIC_KEY, key).kv(QOS_KEY, value.name()).log("Unable to resubscribe to topic");
+    /**
+     * Run re-subscription task in another thread so that the current thread is not blocked by it. The task will keep
+     * retrying until all subscription succeeded or it's canceled by network interruption.
+     *
+     * @param sessionPresent whether the session persisted
+     */
+    private synchronized void resubscribe(boolean sessionPresent) {
+        // Don't bother spin up a thread if subscription is empty
+        if (!subscriptionTopics.isEmpty()) {
+            // If connected without a session, all subscriptions are dropped and need to be resubscribed
+            if (!sessionPresent) {
+                droppedSubscriptionTopics.putAll(subscriptionTopics);
+            }
+            if (!droppedSubscriptionTopics.isEmpty() && (resubscribeFuture == null || resubscribeFuture.isDone())) {
+                resubscribeFuture = executorService.submit(this::resubscribeDroppedTopicsTask);
+            }
+        }
+    }
+
+    private void resubscribeDroppedTopicsTask() {
+        long delayMillis = 0;  // don't delay the first run
+        while (!droppedSubscriptionTopics.isEmpty()) {
+            logger.atDebug().kv("droppedTopics", droppedSubscriptionTopics.keySet()).kv("delayMillis", delayMillis)
+                    .log("Subscribing to dropped topics");
+            ScheduledFuture<?> scheduledFuture = ses.schedule(() -> {
+                for (Map.Entry<String, QualityOfService> entry : droppedSubscriptionTopics.entrySet()) {
+                    CompletableFuture<Integer> subFuture = subscribe(entry.getKey(), entry.getValue());
+                    try {
+                        subFuture.get();
```

Comment:
We can send out all the subscription requests at once. Then use `future.whenComplete` to handle the outcomes from the future. You can use CompleteableFuture.allOf to get you a future that represents all of the other futures completing.

Reply:
Hmm I actually tried that approach and had 2 concerns:
- If some of the future fails,`allOf` will fail, then we seem to have no way to know which ones succeeded and remove those in the next run.
- We probably had to handle interruption [like this](https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/851/commits/165470c714ced3b327efceec1ca08a181baa6f9b#diff-bfd141b30d873984288154d5640bb36b2fec76b936defe54e86977c49312e77eR285-R292), making this a little too complex?

Reply:
You don't need to know which one failed in the allOf, instead you can do something like

```
for topics:
  fut = subscribe(...);
  list.add(fut.whenComplete((i, t) -> {
    // Handle completion logic to remove topic from retry list if it worked.
  }));
allOf(list).get(); // errors can be ignored here since they're handled in the whenComplete, except for InterruptedException
```